src/core/Breadcrumbs/Breadcrumbs.tsx

```diff
@@ -0,0 +1,158 @@
+ * Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+ * See LICENSE.md in the project root for license terms and full copyright notice.
+ *--------------------------------------------------------------------------------------------*/
+import React from 'react';
+import cx from 'classnames';
+import { useTheme } from '../utils/hooks/useTheme';
+import { CommonProps } from '../utils/props';
+import SvgChevronRight from '@itwin/itwinui-icons-react/cjs/icons/ChevronRight';
+import '@itwin/itwinui-css/css/breadcrumbs.css';
+import { useResizeObserver } from '../utils/hooks/useResizeObserver';
+import { useMergedRefs } from '../utils/hooks/useMergedRefs';
+
+export type BreadcrumbsProps = {
+  /**
+   * Index of the currently active breadcrumb.
+   * Defaults to last index in the items array.
+   */
+  currentIndex?: number;
+  /**
+   * Breadcrumb items.
+   */
+  items: JSX.Element[];
+  /**
+   * Specify a custom separator element to show between breadcrumb items.
+   * Defaults to the `SvgChevronRight` icon.
+   */
+  separator?: React.ReactNode;
+} & Omit<CommonProps, 'title'>;
+
+/**
+ * A breadcrumb trail is used as a navigational aid to help users keep track
+ * of their place in the application. It is often placed before a page's main content.
+ *
+ * For accessibility, make sure to provide an aria-label describing the type of navigation.
+ *
+ * @example
+ * <Breadcrumbs
+ *   items={[
+ *     <BreadcrumbItem onClick={() => {}}>Root</BreadcrumbItem>,
+ *     <BreadcrumbItem onClick={() => {}}>Item 1</BreadcrumbItem>,
+ *     <BreadcrumbItem onClick={() => {}}>Item 2</BreadcrumbItem>,
+ *   ]}
+ *   aria-label='Primary'
+ * />
+ />
+ */
+export const Breadcrumbs = React.forwardRef(
+  (props: BreadcrumbsProps, ref: React.RefObject<HTMLElement>) => {
+    const {
+      items,
+      currentIndex = items.length - 1,
+      separator,
+      className,
+      ...rest
+    } = props;
+
+    useTheme();
+
+    const breadcrumbsRef = React.useRef<HTMLElement>(null);
+
+    const [hiddenItems, setHiddenItems] = React.useState<JSX.Element[]>([]);
+    const [postItems, setPostItems] = React.useState(() => items.slice(1));
+    const overflowBreakpoints = React.useRef<number[]>([]);
+
+    const [breadcrumbsWidth, setBreadcrumbsWith] = React.useState<number>(0);
+    const updateBreadcrumbsWidth = React.useCallback(
+      ({ width }) => setBreadcrumbsWith(width),
+      [],
+    );
+    const [resizeRef] = useResizeObserver(updateBreadcrumbsWidth);
+    const refs = useMergedRefs(breadcrumbsRef, resizeRef, ref);
+
+    React.useLayoutEffect(() => {
+      if (!breadcrumbsRef.current) {
+        return;
+      }
+
+      // hide items when there's no space available
+      if (
+        breadcrumbsRef.current.offsetWidth < breadcrumbsRef.current.scrollWidth
+      ) {
+        if (postItems.length > 0) {
+          setHiddenItems([...hiddenItems, postItems.shift()!]); // eslint-disable-line @typescript-eslint/no-non-null-assertion
```

Comment:
`shift`, `pop` mutate existing array and it is considered a bad practice. Also, you might need to use callback setter here to avoid random bugs when resizing and it is set many times.

Reply:
Changed whole code so this is gone 💥